### DISCUSSION TITLE
Wrong city is detected

### DISCUSSION THREAD

**Discussion by golibali** [immich-app/immich#12641] [Q&A] [answered]
### The bug

I live in the countryside of Budapest, in a small town. All the photos what I take in the town is registered by Immich to the town next to us ( there is a hill between us) I checked some of the images coordinates and they are in the town (eg: 47,5767277777778 N, 18,8796 E) and is registered as it is in the Pilisszetivan town.

### The OS that Immich Server is running on

Ubuntu 22.04

### Version of Immich Server

v1.115.0

### Version of Immich Mobile App

1.115.0 build.175

### Platform with the issue

- [X] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
Same as recommended here: https://immich.app/docs/install/docker-compose/
```

### Your .env content

```Shell
Same as recommended here: https://immich.app/docs/install/docker-compose/
```

### Reproduction steps

1. Take an image in town
2. Mobile app uploads
3. Immich server tag with wrong town label
...

### Additional information

I found this ticket: https://github.com/immich-app/immich/issues/4745
I this ticket the https://www.geonames.org/ was mentioned as Immich gets info from there. I checked the wrong "Plilisszentiván" town and the rectengle not overlaps with the geo locations of my images. The right town name would be "Nagykovácsi". There are 3 of them on this page, but they are almost covering the same place. Could this be the root issue, that there are multiple entities for one town in the geonames database?

**Comment by SanoKriss** [comment]
I have same problem with another location 48,1945059722222 17,053632 detected as Devínska Nová Ves and should be Lamač

**Comment by mmomjian** [comment]
The way we handle geo boundaries will be changed in a future release. Currently this is an expected behavior. Thank you

**Reply by golibali** [reply]
Is there a plan date when will be released that future release? Also is there some kind of override instead of editing every image until the fix version is released?

**Reply by mmomjian** [reply]
There is no ETA/override right now, we don’t really do ETAs.

**Reply by SanoKriss** [reply]
Well I try test it.
This [http://api.geonames.org/findNearby?lat=48.1948&lng=17.0526&username=test01](url) return 
`<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<geonames>
    <geoname>
        <toponymName>Lamač</toponymName>
        <name>Lamač</name>
        <lat>48.19298</lat>
        <lng>17.05078</lng>
        <geonameId>3059163</geonameId>
        <countryCode>SK</countryCode>
        <countryName>Slovakia</countryName>
        <fcl>P</fcl>
        <fcode>PPL</fcode>
        <distance>0.24299</distance>
    </geoname>
</geonames>`
but Immich show Devínska Nová Ves, it is definitely  **BUG** 
![screenshot](https://github.com/user-attachments/assets/c81b047a-47bc-4da1-ab8b-252c8e7b95b0)

**Reply by bo0tzz** [reply]
@SanoKriss the geonames api is not representative of how things are expected to work in Immich. It's the same data but a different query method.

**Reply by SanoKriss** [reply]
So, tell me please, which method is used? Thanks

**Reply by golibali** [reply]
Also why do you close this issue as the bug is valid, I mean it tags images with wrong location.

**Reply by kvalev** [reply]
Immich uses the data from the cities500.txt file. So your picture will be geo located in the closest village with a population of more than 500 (within some radius). You can check the geonames data and correct it if needed.

**Reply by golibali** [reply, marked as answer]
@kvalev Thanks, that is the issue here. The "Nagykovácsi" town geo location in the cities500.txt is way too wrong, it's more than 12 km far from the real location of the town.
I checked in this, just for the record: https://download.geonames.org/export/dump/cities500.zip

**Reply by zackpollard** [reply]
![image](https://github.com/user-attachments/assets/163fb289-eca7-4feb-ad47-fe5c160912bb)
This is where it appears on geonames, which does appear to be wrong. You can correct this by going to geonames.org and changing the location of the populated place marker :slightly_smiling_face:

**Reply by SanoKriss** [reply]
> Immich uses the data from the cities500.txt file. So your picture will be geo located in the closest village with a population of more than 500 (within some radius). You can check the geonames data and correct it if needed.

Lamač is part of Bratislava populated with more than 7000 people, but is not in that file. I will check it in geonames, they have several sources of data, for example in allcountry.zip Lamač is there. This is problem of "dirty" data and I understand using source file (cities500.txt) but this not solved problem (not only my). Photos geolocation is wrong marked to nearest city/village.

**Reply by kvalev** [reply]
You can add the population for Lamac [here](https://www.geonames.org/3059163/lamac.html) and after the next immich release if you refresh your metadata, the location should be corrected.
The immich solution is not ideal, but it is what it is and it probably is gonna take a while until another one is in place, so you have to work with what you got.

**Reply by mertalev** [reply]
We're planning on making changes that will make reverse geocoding more accurate. I believe this will also involve using OSM data instead of GeoNames.

**Reply by SanoKriss** [reply]
> You can add the population for Lamac [here](https://www.geonames.org/3059163/lamac.html) and after the next immich release if you refresh your metadata, the location should be corrected. The immich solution is not ideal, but it is what it is and it probably is gonna take a while until another one is in place, so you have to work with what you got.

I can confirm, that Lamač is in cities500 from today. :) Thanks
It is possible (any way) to force refresh or I need wait for next immich release?

**Reply by zackpollard** [reply]
You'll need to wait for the next Immich release 🙂

**Reply by stibbzy** [reply]
> We're planning on making changes that will make reverse geocoding more accurate. I believe this will also involve using OSM data instead of GeoNames.

@mertalev Would this possibly fix an issue I have with geotagged photos from a trip to Antarctica? Despite being tagged, the location is never displayed in Immich. I'm assuming this is because there are no population centres of more than 500 people anywhere.

Other photos from other locations display their locations ok, it's just the Antarctica pics.

**Reply by mertalev** [reply]
The change will make reverse geocoding check which areas contain a coordinate, instead of finding the nearest populated area. Antarctica would *probably* be included as an area, but I'm not sure. Zack knows more about this than I do.

**Reply by zackpollard** [reply]
> > We're planning on making changes that will make reverse geocoding more accurate. I believe this will also involve using OSM data instead of GeoNames.
> 
> @mertalev Would this possibly fix an issue I have with geotagged photos from a trip to Antarctica? Despite being tagged, the location is never displayed in Immich. I'm assuming this is because there are no population centres of more than 500 people anywhere.
> 
> Other photos from other locations display their locations ok, it's just the Antarctica pics.

There were some changes about a month ago which added a country fallback in the case of no matched data in cities500, I'm not sure if it has a region for Antarctica though. You could test by opening one of your Antarctica photos, clicking the three dots in the top right and click "refresh metadata". Give it a few seconds and then refresh the page and see if it now has a country set 🙂

**Reply by stibbzy** [reply]
@zackpollard I just tried the refresh. Unfortunately, it didn’t work. The places are all around the Antarctic peninsula (Anders Island, Neko island, Vernadski research base etc). Google and Apple Maps correctly identifies these locations. 

I noted something though. If I change the location for one of the pics to the main Antarctic landmass (say, the South Pole), then Immich will provide a simple ‘Antarctica’ as the location. Moving the location back to the peninsula and Immich again shoes no location information.

**Reply by zackpollard** [reply]
> @zackpollard I just tried the refresh. Unfortunately, it didn’t work. The places are all around the Antarctic peninsula (Anders Island, Neko island, Vernadski research base etc). Google and Apple Maps correctly identifies these locations. 
> 
> I noted something though. If I change the location for one of the pics to the main Antarctic landmass (say, the South Pole), then Immich will provide a simple ‘Antarctica’ as the location. Moving the location back to the peninsula and Immich again shoes no location information. 

Yea, so it's probably just the main landmass covered by that. If you give me some coordinates I can see what the new dataset would show. It may not be the best initially but it should be possible to improve it if needed . Not sure when the improvements will land, lots of other things going on at the moment 😛

**Reply by stibbzy** [reply]
@zackpollard Sure, here is Vernadksy: 65°14'44.9"S 64°15'28.1"W (-65.245800, -64.257800). It'd be great to have an idea what the new dataset would do with it! Otherwise, I might just have to set all the pics to being on the main landmass.

**Reply by zackpollard** [reply]
> @zackpollard Sure, here is Vernadksy: 65°14'44.9"S 64°15'28.1"W (-65.245800, -64.257800). It'd be great to have an idea what the new dataset would do with it! Otherwise, I might just have to set all the pics to being on the main landmass.

Looks like there are no results returned for Antarctica in my current export, however as you can see in the OSM data, there are polygons for these islands so we should be able to import them, just may not be in the first pass. https://www.openstreetmap.org/relation/2988475

**Reply by stibbzy** [reply]
@zackpollard No worries. Thanks for looking into it. Should I raise a new issue or just wait out for the time being?

**Reply by markggn** [reply]
> Immich uses the data from the cities500.txt file. So your picture will be geo located in the closest village with a population of more than 500 (within some radius). You can check the geonames data and correct it if needed.

Seems not to be true for my pictures. I live in Stuttgart, Germany (500k). Geonames is correct, while Immich shows names of bordering citys way inside the wrong city. 1, 2 and 3 on the following image show "Fellbach", which is clearly multiple kilometers away.

![grafik](https://github.com/user-attachments/assets/fc2662af-79c2-4092-90c9-8e4fa25deeb1)

**Reply by bo0tzz** [reply]
Fellbach is closer to those points than the center of Stuttgart, no?

**Reply by markggn** [reply]
> Fellbach is closer to those points than the center of Stuttgart, no?
Yes. Oh okay, now I seem to understand. This messes up a lot and is quite useless.

**Reply by zackpollard** [reply]
Yes, we are planning to improve this.

**Reply by tomzorz** [reply]
@zackpollard Not to necro the thread, but is there a timeline on when these improvements are planned? I feel like this is one of those issues that is significantly worse in some parts of the world. I tried looking around but I couldn't find an issue tracking it - I am even considering contributing, maybe it'd be a fun issue to tackle as it seems like a fairly nicely separated thing code-wise.

**Reply by bo0tzz** [reply]
We have ideas for improvement but priorities are elsewhere at the moment. If you want to go down the rabbit hole, there's an extensive discussion on our discord server: https://discord.com/channels/979116623879368755/1247238400675807372

**Reply by zackpollard** [reply]
I want to get back to it, but as @bo0tzz mentioned there are other priorities right now. Most of the investigation into consolidating and querying the data was done. The main stuff now is around infrastructure to get the data ingestion, cleaning and serving process which is probably not something that's easy for you to work on really. The code within the server itself shouldn't be that complex, but given everything else that needs doing I'm not sure if that's really worth picking up on its own.

**Reply by gthrepwood** [reply]
What I thinking, an extract from OSM with keys "https://wiki.openstreetmap.org/wiki/Key:admin_level" would be exactly what needed. Not sure, how large the db will be, but for me, I would definitely would choose that...

**Reply by tomzorz** [reply]
By the looks of that, we'd want to do a separate admin_level export per country to get a more accurate deal going. But also more recently I've been wondering if instead of that, a feature where users could designate point+radius or square areas as "home" or "work" would be easier+better UX.

**Comment by gthrepwood** [comment]
I asked someone to extract the exact boundaries from OSM, and without further optimization, db was ~6GB.

I live in the one side of the capital city, and the place if closer to a small village than the center of the capital.

This is a small workaround: set "Budapest XIX" to the photos, what was 10m radius of my location. Yes, it will overwritten, when immich rescan the exif of old photos.

```sql
UPDATE asset_exif ae
SET city = 'Budapest XIX'
FROM asset a
WHERE ae."assetId" = a.id
  AND (
    1000 * 6371 * acos(
        cos(radians(47.42560464469522)) * cos(radians(ae.latitude)) * cos(radians(ae.longitude) - radians(19.201789505737757)) + 
        sin(radians(47.42560464469522)) * sin(radians(ae.latitude))
    )
  ) <= 10;
```

**Reply by tomzorz** [reply]
Oooh actually this gave me an idea. It'd probably take 30 minutes to vibe code a little sidecar docker image where a web UI lets you pick overrides like above on a map, and then run a cron job occasionally to run the DB modifications. Or we could even use that 6GB boundary export of yours and just try map a precise location to every photo + tag them if they were updated just so we don't re-do the work on subsequent runs.

**Reply by kvalev** [reply]
There already is a proposal for a fix by @marcules without hacky solutions - https://github.com/immich-app/immich/pull/9051#issuecomment-3831064032 The only thing missing is a blessing from the maintainers to go with the fix.

**Reply by gthrepwood** [reply]
As I see, the request pending from Apr 24, 2024 and that is an awful lot of time. It seems to be 'better' that the original cities500 solution, but still questionable, if that would be a trusted improvement.

**Reply by marcules** [reply]
Hey hi, @gthrepwood my proposal is actually only about 2 weeks old, but I made the mistake of commenting it on an old PR instead of this discussion, as, from my point of view, the linked PR actually introduced a regression for anyone taking pictures in the outskirts of cities.

So I'm not sure if you've read the actual proposal further down the comments, because I don't think it's questionable to use the hierarchy of the nearest populated area to determine the "correct" nearest city.

Of course it would be even better to use actual geofencing or even allow the user to designate their own locations as to create "synthetic"/"manual" populated area markers, that would further improve pictures in that region for the user in question; but the proposal is deliberately small in scope as to not impact any plans the maintainers have with real geofencing, so not to impose on their ideas of a "real geofencing implementation" this is as small as possible, but with the desired effect of improving the location data on outskirts.

**Reply by tomzorz** [reply]
there was also a thread somewhere about a better solution being implemented / halfway done already; or maybe it was in discord? not sure. but it wasn't #9051

**Reply by MichelZ** [reply]
I've built exactly this now, it uses overture maps and uses a place's geometry/bounding box, not just a point on the map to calculate the place: 
Docs: https://immich-reversegeo.github.io/
GitHub: https://github.com/immich-reversegeo/immich-reversegeo

**Comment by MichelZ** [comment]
I've built exactly this now, it uses overture maps and uses a place's geometry/bounding box, not just a point on the map to calculate the place: 
Docs: https://immich-reversegeo.github.io/
GitHub: https://github.com/immich-reversegeo/immich-reversegeo

**Reply by gthrepwood** [reply]
I tried @MichelZ  's solution and it's more than enogh as I see. on demand database download, easy to install, works as expected.

I checked OSM's database, how to extract city borders, and it's VERY tricly as China, Greece, Hungary have all use different levels for administrative regions.... I would gladly accept, if someone else done  that work, and harmonise the databases.

**Reply by ClementTsang** [reply]
Ooh, I might try this - I have some photos that are smack dab in the middle of a huge city that it currently decides is a city a few miles over for whatever reason.
